src/decode/ykpt/packet_parser/mod.rs

```diff
@@ -0,0 +1,303 @@
+        // frequently expected packet kinds first.
+        //
+        // OPT: The order below is a rough guess based on what limited traces I've seen. Benchmark
+        // and optimise.
+        match self {
+            Self::Init => &[PacketKind::PSB],
+            Self::Normal => &[
+                PacketKind::ShortTNT,
+                PacketKind::PAD,
+                PacketKind::FUP,
+                PacketKind::TIP,
+                PacketKind::CYC,
+                PacketKind::LongTNT,
+                PacketKind::PSB,
+                PacketKind::MODE,
+                PacketKind::TIPPGE,
+                PacketKind::TIPPGD,
+            ],
+            Self::PSBPlus => &[PacketKind::CBR, PacketKind::PSBEND],
+        }
+    }
+
+    /// Check if we need to transition state after parsing a given packet kind.
+    fn transition(&mut self, pkt_kind: PacketKind) {
+        let new = match (*self, pkt_kind) {
+            (Self::Init, PacketKind::PSB) => Self::PSBPlus,
+            (Self::Normal, PacketKind::PSB) => Self::PSBPlus,
+            (Self::PSBPlus, PacketKind::PSBEND) => Self::Normal,
+            _ => return, // No state transition.
+        };
+        *self = new;
+    }
+}
+
+pub(super) struct PacketParser<'t> {
+    /// The raw bytes of the PT trace we are iterating over.
+    bytes: &'t [u8],
+    /// The parser operates as a state machine. This field keeps track of which state we are in.
+    state: PacketParserState,
+    /// The most recent Target IP (TIP) value that we've seen. This is needed because updated TIP
+    /// values are sometimes compressed using bits from the previous TIP value.
+    prev_tip: usize,
+}
+
+/// Attempt to read the packet of type `$packet` using deku. On success wrap the packet up into the
+/// corresponding discriminant of `Packet`.
+macro_rules! read_to_packet {
+    ($packet: ty, $bits: expr, $discr: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| Ok((r, $discr(p))))
+    };
+}
+
+/// Same as `read_to_packet!`, but with extra logic for dealing with packets which encode a TIP.
+macro_rules! read_to_packet_tip {
+    ($packet: ty, $bits: expr, $discr: expr, $prev_tip: expr) => {
+        <$packet>::read($bits, ()).and_then(|(r, p)| {
+            let ret = if p.needs_prev_tip() {
+                Ok((r, $discr(p, Some($prev_tip))))
+            } else {
+                Ok((r, $discr(p, None)))
+            };
+            ret
+        })
+    };
+}
+
+impl<'t> PacketParser<'t> {
+    pub(super) fn new(bytes: &'t [u8]) -> Self {
+        Self {
+            bytes,
+            state: PacketParserState::Init,
+            prev_tip: 0,
+        }
+    }
+
+    fn mk_parse_error(&mut self) -> HWTracerError {
+        let mut s = String::new();
+        write!(
+            &mut s,
+            "In state '{:?}', failed to parse packet: ",
+            self.state
+        )
+        .unwrap();
```

Comment:
Isn't this function just an odd way of saying `format!("In state {:?}, failed to parse packet {}", self.state, self.byte_stream_str(...))`?

Reply:
It used to be called from many places, but I see there is now only one call site. We can inline it, yes.

Reply:
Fixed in [3567abf](https://github.com/ykjit/hwtracer/pull/105/commits/3567abfb722b6106c105a07bc9a0b4ceda47d3a4)